File: app/src/main/java/com/esbati/keivan/persiancalendar/Utils/AndroidUtilities.java
package com.esbati.keivan.persiancalendar.Utils;

import android.annotation.TargetApi;
import android.app.ActivityManager;
import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.content.res.Configuration;
import android.graphics.Point;
import android.graphics.Typeface;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Display;
import android.view.View;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;

import com.esbati.keivan.persiancalendar.Components.ApplicationController;

import java.util.Hashtable;
import java.util.List;

public class AndroidUtilities {

    private static float density;
    private static final Hashtable<String, Typeface> typefaceCache = new Hashtable<>();

    static {
        density = ApplicationController.getContext().getResources().getDisplayMetrics().density;
    }

    public static int dp(float value) {
        if (value == 0) {
            return 0;
        }
        return (int) Math.ceil(density * value);
    }

    public static Typeface getTypeface(String assetPath) {
        synchronized (typefaceCache) {
            if (!typefaceCache.containsKey(assetPath)) {
                try {
                    Typeface t = Typeface.createFromAsset(ApplicationController.getContext().getAssets(), assetPath);
                    typefaceCache.put(assetPath, t);
                } catch (Exception e) {
                    Log.e("Typefaces", "Could not get typeface '" + assetPath + "' because " + e.getMessage());
                    return null;
                }
            }
            return typefaceCache.get(assetPath);
        }
    }

    public static void showSoftKeyboard(View view){
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) ApplicationController.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.showSoftInput(view, InputMethodManager.SHOW_IMPLICIT);
        }
    }

    public static void hideSoftKeyboard(View view){
        if (view != null) {
            InputMethodManager imm = (InputMethodManager)ApplicationController.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    public static boolean isServiceRunning(Class<?> serviceClass) {
        ActivityManager manager = (ActivityManager) ApplicationController.getContext().getSystemService(Context.ACTIVITY_SERVICE);
        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            if (serviceClass.getName().equals(service.service.getClassName())) {
                return true;
            }
        }
        return false;
    }

    @TargetApi(17)
    public static void showRTLDialog(AlertDialog dialog){
        dialog.show();
        //Set Title Gravity
        final int alertTitle = ApplicationController.getContext().getResources().getIdentifier("alertTitle", "id", "android");
        TextView messageText = dialog.findViewById(alertTitle);
        messageText.setLayoutDirection(View.LAYOUT_DIRECTION_RTL);
    }
}

File: app/src/main/java/com/esbati/keivan/persiancalendar/utils/AndroidUtilities.kt
package com.esbati.keivan.persiancalendar.utils

import android.content.Context
import android.os.Build
import android.support.annotation.ColorRes
import android.support.annotation.FontRes
import android.support.annotation.StringRes
import android.support.v4.content.ContextCompat
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.app.AlertDialog
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import android.widget.Toast
import com.esbati.keivan.persiancalendar.R
import com.esbati.keivan.persiancalendar.components.ApplicationController


private val density = ApplicationController.getContext().resources.displayMetrics.density

fun dp(value: Float): Int {
    return if (value == 0f) 0 else Math.ceil((density * value).toDouble()).toInt()
}

fun Int.toDp(): Int {
    return if (this == 0) 0 else Math.ceil((density * this).toDouble()).toInt()
}

fun Float.toDp(): Int {
    return if (this == 0f) 0 else Math.ceil((density * this).toDouble()).toInt()
}

fun View.showSoftKeyboard() {
    val imm = this.context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
    imm.showSoftInput(this, InputMethodManager.SHOW_IMPLICIT)
}

fun View.hideSoftKeyboard() {
    val imm = this.context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
    imm.hideSoftInputFromWindow(this.windowToken, 0)
}

fun AlertDialog.Builder.showThemedDialog(): AlertDialog {
    val dialog = this.create()
    dialog.showThemedDialog()
    return dialog
}

fun AlertDialog.showThemedDialog() {
    show()
    applyDefaultTheme()
}

fun AlertDialog.applyDefaultTheme(): AlertDialog {
    if (Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN)
        findViewById<TextView>(R.id.alertTitle)?.layoutDirection = View.LAYOUT_DIRECTION_RTL

    applyFont()
    return this
}

fun AlertDialog.applyFont(@FontRes title: Int = R.font.iransans_fa_num_bold,
                          @FontRes message: Int = R.font.iran_sans,
                          @FontRes buttons: Int = R.font.iran_sans): AlertDialog {

    findViewById<TextView>(R.id.alertTitle)?.typeface = ResourcesCompat.getFont(context, title)
    findViewById<TextView>(android.R.id.message)?.typeface = ResourcesCompat.getFont(context, message)
    findViewById<TextView>(android.R.id.button1)?.typeface = ResourcesCompat.getFont(context, buttons)
    findViewById<TextView>(android.R.id.button2)?.typeface = ResourcesCompat.getFont(context, buttons)
    findViewById<TextView>(android.R.id.button3)?.typeface = ResourcesCompat.getFont(context, buttons)

    return this
}

fun getColor(@ColorRes value: Int): Int {
    return ContextCompat.getColor(ApplicationController.getContext(), value)
}

fun showToast(@StringRes value: Int, duration: Int = Toast.LENGTH_SHORT) {
    Toast.makeText(ApplicationController.getContext(), value, duration).show()
}



